package.json

```diff
@@ -36,6 +36,7 @@
   },
   "devDependencies": {
     "@babel/cli": "^7.17.6",
+    "@babel/core": "^7.18.9",
```

Comment:
What are the relative pros/cons of explicitly adding missing peerDeps like this compared to setting `auto-install-peers=true`?

Naively I would have assumed that having peers autoinstall would be less of a maintenance burden long-term, since we don't actually care about what `@babel/core` version is available for our own purposes; we just want the peer requirement (whatever it happens to be) to be satisfied. You've got more real-world experience dealing with pnpm than me, though, so I'm curious about the conventional wisdom here 🙂

Reply:
I only have fears with auto installing things from back when i was using parcel.

I do like being explicit, and i think doing so reveals than we have upstream problems to fix in the ecosystem. In this case, specifically, i think (need to verify) we need to add @babel/core to all packages interacting with babel. There is a list in the commit.

I should probably add some more 'notes' to the package.json about this 🤔

Reply:
> I only have fears with auto installing things from back when i was using parcel.

What went wrong with auto peerDeps and Parcel?

> we need to add @babel/core to all packages interacting with babel

Why is that? It looks this is mostly due to dependency chains that end at various `@babel/plugin-*` packages, which each declare a peerDep on `@babel/core`.

Is it the responsibility of the thing that relies on the plugin package to include `@babel/core` as a dependency? If so, why? If not, whose responsibility is it?

Reply:
It's been ages since parcel so i don't totally remember. Not a good reason for fear, but 🙃

For the peers, i think @ef4 's explanation is better than i could do https://github.com/ember-cli/ember-cli/pull/9934

Reply:
Which, after rereading those comments, i think i'm confused about what the plan should be

Reply:
I think he's saying each package in the chain should recursively re-declare its dependencies' `peerDependencies` as its own, ultimately forcing the host to declare an explicit dependency on them.

If I'm reading that right (which I may not be!), I'm not sure that makes sense to me—peerDeps already effectively percolate up to the host regardless of whether the intermediary dependencies declare them or not. Having them redeclare their peerDeps seems to me like it just adds another point of synchronization that authors need to maintain.

Concretely: suppose my library `dans-fun-lib` depends on `neat-tools@^1.2.3`, which has a peerDep on `neat-core@^2.4.0`. Regardless of whether `dans-fun-lib` declares its own peerDep on `neat-core`, the host will need to ensure `neat-core` is available at an appropriate version or else a warning or error will be emitted by the package manager.

If I _do_ redeclare the peerDep, and then `neat-tools` updates its range to accept `neat-core@^2.4.0 || ^3.0.0`, now I need to make a new release of my library that widens its own acceptable range in order not to block apps from using `neat-core@3.0.0`, _even though my library doesn't actually care about `neat-core` at all_.

Reply:
Worth noting: all of the above is true whether the host declares its own dependency on `neat-core` or just lets their package manager deal with it for them (as npm@7+ does, or pnpm with `auto-install-peers`)

Reply:
A problem is that `dans-fun-lib` can communicate two different things:

1. "I don't care what copy of neat-core gets used by my copy of neat-tools". This is a valid thing to communicate, when it would not be a bug for multiple copies of neat-core to coexist.
2. "I want to use the same copy of neat-core that my consumer(s) are using". This is what you need to communicate when there is shared global state involved such that it would be a bug for multiple copies of neat-core to be involved.

To communicate the second thing, `dans-fun-lib` needs it own peer dependency. Otherwise, a valid solution is for the package manager is to install lots of copies of `neat-core`:

```
the-app/node_modules/dans-fun-lib/node_modules/neat-core
the-app/node_modules/something-else/node_modules/neat-core
the-app/node_modules/neat-core
```

Consider what should happen if `dans-fun-lib` and `something-else` end up needing incompatible versions of `neat-core`. In scenario 1 above, it would be best if the package manager solved that automatically by installing both versions. In scenario 2 above, it should error because there's no valid solution and the user will need to upgrade or downgrade some things to get everybody onto a compatible version.

I can't say that I've tested all the major NPM clients to see which way they happen to go today in these scenarios. But since both scenarios are valid things for a package to want, it seems better to me to explicitly distinguish between them.

Reply:
In cases where it's truly a problem to have two copies of a library present, I agree that you want that to manifest to the user as a dependency resolution error. I don't think `@babel/core` (which is what kicked us off here) is an example of a package like that, though, unless I'm missing something key there?

Reply:
this conversation is mostly handled by the ember-cli-babel@8 efforts, as well as the pnpm support in ember-cli RFC.
Will resolve for now -- further conversations should probably be on the pnpm RFC.